src/backend/distributed/planner/relation_restriction_equivalence.c

```diff
@@ -2062,25 +2063,47 @@ ContextCoversJoinRestriction(JoinRestrictionContext *joinRestrictionContext,
 			continue;
 		}
 
+		List *joinRestrictInfoListInTest =
+			joinRestrictionInTest->joinRestrictInfoList;
+		bool hasJoinRestriction = list_length(joinRestrictInfoListInTest) != 0;
+		bool hasOnlyInnerJoin = joinRestrictionContext->hasOnlyInnerJoin;
+		if (!hasOnlyInnerJoin && !hasJoinRestriction)
+		{
+			/*
+			 * If join doesn't have a restriction (e.g., ON (true)) and planner
+			 * is aware of at least one non-inner JOIN (e.g., outer/semi joins),
+			 * we should not eliminiate joinRestrictionInTest. It can still be
+			 * useful for detecting not supported outer-join checks even if it
+			 * doesn't help for colocation checks.
+			 */
+			continue;
+		}
+
 		/*
 		 * We check whether the restrictions in joinRestrictionInTest is a subset
 		 * of the restrictions in joinRestrictionInContext in the sense that all the
 		 * restrictions in the latter already exists in the former.
-		 *
-		 * Also, note that list_difference() returns a list that contains all the
-		 * cells in joinRestrictInfoList that are not in inputJoinRestrictInfoList.
-		 * Finally, each element in these lists is a pointer to RestrictInfo
-		 * structure, where equal() function is implemented for the struct.
 		 */
 		List *joinRestrictInfoListInContext =
 			joinRestrictionInContext->joinRestrictInfoList;
-		List *joinRestrictInfoListInTest =
-			joinRestrictionInTest->joinRestrictInfoList;
-		if (LeftListIsSubset(joinRestrictInfoListInContext, joinRestrictInfoListInTest))
+		if (LeftListIsSubset(joinRestrictInfoListInTest, joinRestrictInfoListInContext))
 		{
 			return true;
 		}
 	}
 
 	return false;
 }
+
+
+/*
```

Comment:
I think this function is not very much necessary. The name is hard to follow etc. Can't we inline it back?

Reply:
I'm okay with removing it, will do